### DISCUSSION TITLE
bytearrays and memoryview

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#14071] [Using MicroPython]
I'm working my way through stuff the rest of the forum members seem to have got sorted back in 2016 so please excuse my late to the party but memoryview has me confused.
I can change a byte array without memory view
```
>>> a=bytearray(b'012')
>>> a[1:2]=b'y'
>>> a
bytearray(b'0y2')
```
or with it
```
>>> a=bytearray(b'012')
>>> mv=memoryview(a)
>>> mv[1:2]=b'y'
>>> a
bytearray(b'0y2')
```
since both alter the original bytearray I don't understand the advantage of the memoryview method? Does the memoryview technique offer some advantage if the bytearray is large (20k)?

**Comment by jimmo** [comment]
@kjm1102 This scenario (slice assignment) is basically the one case where memoryview doesn't help.

Memoryview allows you to make read/write slices of some original buffer, _without copying the contents of the buffer_.

Consider instead:

```py
def foo(buf):
  print(buf.hex())

b = bytes(range(10))
foo(b[1:4])
m = memoryview(b)
foo(m[1:4])
```

Both print the same thing, but the first call to `foo` causes a copy of the bytes to be created, the second does not.

Maybe a better example:

```py
def foo(buf):
  buf[2] = 0
  print(buf.hex())

b = bytearray(range(10))
foo(b[1:4])
print(b.hex())
m = memoryview(b)
foo(m[1:4])
print(b.hex())
```

It prints

```
010200  <-- copy changed  (03 is now 00)
00010203040506070809    <-- original unchanged  (03 is still there)
010200   <-- modifying memoryview  (03 is now 00)
00010200040506070809    <-- original changed  (03 is now 00)
```
